### ISSUE TITLE
Block joints

### ISSUE THREAD

**Issue by thscharler** [ratatui/ratatui#1468]
Not an issue but a proposal:

[ratatui_block](https://thscharler.github.io/ratatui-block/doc/ratatui_block/index.html)

Not all combinations are covered yet, but that's just a bit more work.

**Comment by orhun** [comment]
Interesting, can you give us some information about your use case and how this looks like?

**Comment by thscharler** [comment]
![Image](https://github.com/user-attachments/assets/95e3422b-ebdf-4d08-bb79-9e2693bfee9f)

**Comment by thscharler** [comment]
![Image](https://github.com/user-attachments/assets/2d7faa33-4fc2-4c88-b63c-7b14a37107fe)

**Comment by thscharler** [comment]
* Joint would be a building part for block-block connections. 

A list of such joints would be added to `Block` and rendered with it. 
The point for adding them to block is the titles. It's difficult to have both
without support from Block. 

A secondary use case for Joint is to draw joints inwards as connectors for
subdivisions within the widget. I currently have a few hundred lines of code in [Split](https://docs.rs/rat-widget/latest/rat_widget/splitter/index.html) just for that.


* `fn create_border(layout: Rc<[Rect]>, n: usize, border: BorderType) -> NewBlock<'static>`

Creates a list of such joints and creates a Block for the selected area, 
considering all connections to the other areas in the list. 

This here is currently limited to one BorderType, but that's more of a convenience.

I'm not sure where to put this one. Maybe as a associated function in Layout? Would probably make it easier to find it.

**Comment by thscharler** [comment]
I'm currently working on the look of side by side blocks with QuadrantInside borders. They still need special casing in create_border() and look a bit clunky. 

I may add fg/bg colors to Joint to have a nice side by side look for those, but I'm not done with that yet. 

Apart from that there are still some cases that render 'X', but this is just a bit of detail work.

**Comment by thscharler** [comment]
Apart from that there are a few border::Set's that are not covered by BlockType. They should probably be added too. Maybe extend the Blocktype enum? When I started my first thoughts where: Ok, BlockType, that's it. And only later I accidently stumbled upon those. So maybe adding them would help visibility.

**Comment by joshka** [comment]
Neat! This feels like the continuation of where we left off in getting the borders happening. We added overlap in #1398 to allow for something like this to happen. There's some docs in https://github.com/ratatui/ratatui/issues/605, and some discussion about this in https://discord.com/channels/1070692720437383208/1291063903987433503

From an API design perspective, I had previously thought (in passing) about making the border parts into their own widget rather than the joints themselves being the concrete part, so that we could independently draw just border parts when needed (with the block widget being the first consumer of this). What do you think about that idea?

From a visualization of solution, I'd suggest displaying all the combinations in a static form rather than animated (it's difficult e.g. to see how cross intersections look). I like how the colors help indicate which border is which.

The other thing to consider is how to render the combination of block types. Particularly the main line drawing types - plain, rounded, double, dotted, solid, etc. There's a variety of intersecting types in Unicode for these, which would be great to see. And besides the built-in combos, this probably needs some way for users to configure their own intersection types. (q: does intersection sound better than joint for this?)

A quick code thing - try to avoid abbreviations in type names when possible as it's difficult to guess when something will be abbreviated vs not (and which abbreviation will be used when it is). (JointPos)

I had a brief skim over the docs and the types / function names didn't immediately make sense to me, so perhaps some narrative would add value. I'd suggest trying to align things as much as possible with the unicode naming / Ratatui symbols naming, or documenting how to align the two ideas.

Definitely keen to see where this goes, and likely would want to see it inside Ratatui rather than external if possible as solving overlapping borders is definitely desired by uses.

**Comment by thscharler** [comment]
> Neat! This feels like the continuation of where we left off in getting the borders happening. We added overlap in [#1398](https://github.com/ratatui/ratatui/pull/1398) to allow for something like this to happen. There's some docs in [#605](https://github.com/ratatui/ratatui/issues/605), and some discussion about this in https://discord.com/channels/1070692720437383208/1291063903987433503

> 
> From an API design perspective, I had previously thought (in passing) about making the border parts into their own widget rather than the joints themselves being the concrete part, so that we could independently draw just border parts when needed (with the block widget being the first consumer of this). What do you think about that idea?

No, I have 2 involved borders of 6 types currently , 4 sides of the area, 3 types of markings (excluding manual), and 7 positions along each side which gives 3024 combinations. For QuadrantInside I needed an extra mirrored flag and for Thick there is a third border involved in the corners. Even if a lot of those end up being the same codepoint you need some way to adress them. 

Behind Joint lies a massive match that is currently 1000 lines long. It's not yet finished. I already reduced it with the use of '_' .

You could theoretically have a widget per codepoint or so, but that still wouldn't solve the combinatorics. 

I've checked the generated assembly, it seems compilers are good at this kind of thing, it's about the same lines in assembly too. 


> 
> From a visualization of solution, I'd suggest displaying all the combinations in a static form rather than animated (it's difficult e.g. to see how cross intersections look). I like how the colors help indicate which border is which.

The video was just for quick illustration. The examples of ratatui block have two examples where you can step through everything. 

> 
> The other thing to consider is how to render the combination of block types. Particularly the main line drawing types - plain, rounded, double, dotted, solid, etc. There's a variety of intersecting types in Unicode for these, which would be great to see. And besides the built-in combos, this probably needs some way for users to configure their own intersection types. (q: does intersection sound better than joint for this?)

For the naming: I'm not a native english speaker, and my translation told me that joint is adequate for a weld point, a connection, or some in-between parts holding things together. So I found it fitting. 

So far I've coverd all combinations of Thick/Plain borders. Plain/Double is fine too, except there are not enough codepoints to cover everything. Rounded is the same a plain. 

And there is a manual mark where you can set everything you want. 


> 
> A quick code thing - try to avoid abbreviations in type names when possible as it's difficult to guess when something will be abbreviated vs not (and which abbreviation will be used when it is). (JointPos)

Fine

> 
> I had a brief skim over the docs and the types / function names didn't immediately make sense to me, so perhaps some narrative would add value. I'd suggest trying to align things as much as possible with the unicode naming / Ratatui symbols naming, or documenting how to align the two ideas.

I have indeed some concept behind the naming, but as said I might have used the wrong translation. I have no strong opinions on naming, so this can be sorted out. 

And there is currently an old version which was insufficient and a new version mixed. I still use the old one for reference and will cleanup when I'm done. 

Anyway, the primary idea is to only consider one side of the Rect (=JointSide). From there you can set a marker (=JointMark) where some connection happens. Which point along the side is given by a position from the start of the line (=JointPos{ition}). There is of course some overlap, e.g. a CrossStart connection on the Left side of the area will give the same as a CrossStart on the top side. But this greatly simplifies creating the connection points for two areas as I only have to consider the side both areas are touching, and don't need to 'think around the corner' too. 

The current implementation for connecting the areas is only 250 loc and a single run over all the areas for one border. This ends in O(n^2) if we do it for all areas in the set, but I think you can't get any better. In the end its only 40 comparisons per area, so this shouldn't be a real problem. 


> 
> Definitely keen to see where this goes, and likely would want to see it inside Ratatui rather than external if possible as solving overlapping borders is definitely desired by uses.

Yes, I definitely would like to see this as part. With the titles on the Block it would be no fun doing it otherwise.

**Comment by joshka** [comment]
Yeah go for it. I might be a bit delayed in looking deeply at things over the next couple of weeks as I have a lot on my plate, so please don't take a lack of response as a lack of interest in this.

What I meant about Border being a widget was similar (but maybe not exactly the same) as what you're saying with JointSide. I meant that it might be useful to treat a bordered Block as having 4 border widgets (Top, Left, Bottom, Right) and to then render each border in a way that generates the right intersecting points with each border are done right.

For the logic in plain_joint, it may be worth considering naming things to align closely with the unicode naming E.g. `╢` is "Vertical double and left single" (where your code has it categorized in the matching table as (Out, Left, Pos(_), Plain). (Obviously this doesn't work for the quadrant / custom types, but it should for most of the box drawing types). This might allow you to generalize the matching logic rather than having to work the combinations. But it's also possible I'm not looking at this from the right perspective as you've gone much deeper than I have already on this.

**Comment by thscharler** [comment]
It will take some time to check all combinations anyway. I will use a non pub copy where I need it now, and replace later. No problem.

The exact layout of the types is still in flux. I think I can remove JointMark and just do with JointPosition and JointSide. This would be similar to what you are describing. 

And, if this can describe the joints it would be no problem to add "just a plain line", "just the start corner", ... and use this for normal block rendering too. We could hide the gory details per BorderType and provide them as a trait impl per BorderType. That trait might solve the user-defined border too. ... Just some thoughts ... 

For the naming: That shows me those unicode guys also have the (hard) problem of naming things. Their names are just visual descriptions, which is fine for unicode but not very helpful when coming at the problem from a different angle. 

A ╢ "Vertical double and left single" can be 
* left side of an area with a connector pointing outwards
* right side of an area with a connector pointing inwards

And I really like the approach that I 'want to overlap the left side of this with the right side of that'.

---

I will look at those too.

**Comment by thscharler** [comment]
term-table-rs: uses something similar to border::Set

Ok, but with the same limitations we already have. 

comfy-table: The same but encoding the positions in one string. 

Ok for them, but hard to read. 

cli-table: They do it linewise, therefore even less parameters. 


So. Tables are a starting reference, but they only have to look inwards so that reduces their problem significantly. I don't want to go with a string with magical indexes. A struct approach is what we already have, and there are too many combinations and no good names. Still a possibility; maybe if we view this as a per side struct?

Or stick with the algorithmic/jump table...

I will look into the 'struct per side' too.

**Comment by thscharler** [comment]
Drew up a [v3](https://thscharler.github.io/ratatui-block/doc/ratatui_block/v3/index.html) for the structs/enums. 

Maybe you can give this a quick look.

**Comment by thscharler** [comment]
> Drew up a [v3](https://thscharler.github.io/ratatui-block/doc/ratatui_block/v3/index.html) for the structs/enums.
> 
> Maybe you can give this a quick look.

Made a few more changes to the design. I think I hit a sweet spot now.

**Comment by thscharler** [comment]
... one last thought for today ...

Using [BlockBorder](https://thscharler.github.io/ratatui-block/doc/ratatui_block/v3/struct.BlockBorder.html) to render all of the border would provide a smooth transition too. 

All current mechanics can stay as they are, we only need to add BlockBorder as an extra. 

If a BlockBorder is present use it for rendering otherwise fall back to old styles.

**Comment by thscharler** [comment]
I think have this down to a usable set of features. 

Could you take a look when you have time...

Afterwards, we need to talk how to integrate any of this into Block.

**Comment by joshka** [comment]
Definitely. I've put this on the back burner a little to focus on #1388 (which is important for the next release as it solves release incompatibility issues when we release changes to widgets)

**Comment by joshka** [comment]
This got mostly implemented in https://github.com/ratatui/ratatui/pull/1729 and then https://github.com/ratatui/ratatui/pull/1874
The implementation approach was a little different in that instead of merging blocks, the merge happens at the cell / symbol level in the buffer. 

There are some parts that are not quite there yet, so leaving this open to track them:
- [x] merging titles on blocks. https://github.com/ratatui/ratatui/pull/1899 has a demonstration of this problem. I'd like this to be in 0.30
- [ ] merging quadrant / half block items (demo of this is above). We should add these. This can slip to 0.31 if needed (though it's possible that this is fairly straightforward).

**Comment by j-g00da** [comment]
#1939 (fix for issue in #1899) that was planned for 0.30 got implemented, only remaining part left is quadrant/half-block merging. 
I will move this issue to 0.31 (we can probably release 0.31 sooner than later, as we can do it without bumping core).
